proposals/21-11-transformer-annotations.md

```diff
@@ -0,0 +1,253 @@
+
+`type` has two legal values, custom or builtin, depending on whether the transformer is built into kustomize or is defined by the user.
+
+#### Remote transformers
+For transformers defined in a remote base, kustomize will an annotation like the following:
+
+```yaml
+config.kubernetes.io/transformations: |
+  - kind: NamespaceTransformer
+    path: examples/multibases/base/kustomization.yaml
+    repo: https://github.com/kubernetes-sigs/kustomize
+    ref: v1.0.6 
+    type: builtin
+```
+
+The only difference between the remote transformer and the local one is that the remote transformer will have `repo` and `ref` fields to store where the 
+transformer originated from. 
+ 
+A resource that has been through multiple transformers through several layers may be outputted by kustomize build with an annotation like the following:
+
+```yaml
+config.kubernetes.io/transformations: |
+  - kind: PrefixSuffixTransformer
+    path: ../overlays/kustomization.yaml
+    type: builtin
+  - kind: InflateHelmChart
+    path: inflatehelmchart.yaml
+    type: custom
+  - kind: NamespaceTransformer
+    path: examples/multibases/base/kustomization.yaml
+    repo: https://github.com/kubernetes-sigs/kustomize
+    ref: v1.0.6 
+    type: builtin
+```
+
+### Kustomize edit
+
+We will want to provide convenient commands for users to edit the `buildMetadata` field in their kustomization.
+We should support the following commands:
+
+`kustomize edit add buildMetadata originAnnotations`
+
+`kustomize edit add buildMetadata transformerAnnotations`
+
+`kustomize edit add buildMetadata *`
+
+`kustomize edit remove buildMetadata originAnnotaitons`
+
+`kustomize edit remove buildMetadata transformerAnnotations`
+
+`kustomize edit remove buildMetadata *`
+
+The wildcard match `*` should add/remove all possible options for `buildMetadata`.
+ 
+
+### User Stories
+
+#### Story 1
+
+I am running kustomize build on some configuration that has already been created. This configuration has a large number
+of resources and patches through multiple overlays in different directories. After running `kustomize build`, I get a 
+stream of 100+ resources, but I notice that a few of them have a typo, likely originating from the same base or patch 
+file. However, because my directory structure is large and complex, it is difficult to narrow down which base these
+resources originated from and which patches have been applied to it. In the top-level kustomization file, I can add a
+new `buildMetadata` field:
+
+```yaml
+buildMetadata: [originAnnotations, transformerAnnotations]
+```
+
+The output of `kustomize build` will contain patched resources with annotations similar to the following:
+```
+annotations:
+  config.kubernetes.io/origin: |
+    path: deployment.yaml
+  config.kubernetes.io/transformations:
+  - kind: PatchTransformer
+    path: ../base/kustomization.yaml
+    type: builtin
+  - kind: PatchStrategicMergeTransformer
+    path: ../dev/patch.yaml
```

Comment:
Are these paths relative to the transformer's parent Kustomization? Might it be hard to figure out where that is, in a large config set as in this example?

Reply:
All paths would be relative to the top-level kustomization file (similar to the `originAnnotation` path field), i.e. the directory upon which `kustomize build` is invoked. 

I saw in another one of your comments that you suggested using the absolute path; while I can see in some scenarios how that might be more clear, using the absolute path would make the output unstable. If we use absolute paths and someone were to clone a kustomize directory to a different folder, they would get different annotations. Having the paths relative to the top-level kustomization file will stabilize the output in these cases.

Reply:
The absolute path suggestion was in the context of the hypothetical that this feature could be turned into a dedicated debugging tool instead of annotations that get persisted. I 100% agree that we cannot put absolute paths in resource annotations.